crates/core/src/routing_info/mod.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::pin::pin;
+use std::sync::Arc;
+
+use anyhow::Context;
+use arc_swap::ArcSwap;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::MissedTickBehavior;
+use tracing::{debug, info, trace};
+use xxhash_rust::xxh3::Xxh3Builder;
+
+use restate_types::cluster_controller::SchedulingPlan;
+use restate_types::config::Configuration;
+use restate_types::identifiers::PartitionId;
+use restate_types::metadata_store::keys::SCHEDULING_PLAN_KEY;
+use restate_types::NodeId;
+
+use crate::metadata_store::MetadataStoreClient;
+use crate::{
+    cancellation_watcher, task_center, ShutdownError, TargetVersion, TaskCenter, TaskHandle,
+    TaskId, TaskKind,
+};
+
+pub type CommandSender = mpsc::Sender<Command>;
+pub type CommandReceiver = mpsc::Receiver<Command>;
+
+pub enum Command {
+    /// Request an on-demand refresh of routing information from the authoritative source.
+    SyncRoutingInformation(TargetVersion),
+}
+
+/// Holds a view of the known partition-to-node mappings. Use it to discover which node(s) to route
+/// requests to for a given partition. Compared to the partition table, this view is more dynamic as
+/// it changes based on cluster nodes' operational status. This handle can be cheaply cloned.
+#[derive(Clone)]
+pub struct PartitionRouting {
+    sender: CommandSender,
+    /// A mapping of partition IDs to node IDs that are believed to be authoritative for that serving requests.
+    partition_to_node_mappings: Arc<ArcSwap<PartitionToNodesRoutingTable>>,
+}
+
+impl PartitionRouting {
+    /// Look up a suitable node to answer requests for the given partition.
+    pub fn get_node_by_partition(&self, partition_id: PartitionId) -> Option<NodeId> {
+        self.partition_to_node_mappings
+            .load()
+            .inner
+            .get(&partition_id)
+            .copied()
+    }
+
+    /// Request a refresh without waiting for it to complete. This is useful when the caller
+    /// discovers via some other mechanism that the local view might be outdated - for example, when
+    /// a request to a node previously returned by `get_node_by_partition` fails with a response
+    /// that indicates that this routing information is no longer valid.
```

Comment:
How will a user of `PartitionRouting` use it if they encounter invalid routing information?

Reply:
Just call it? :-) It's super dumb right now, just a fire and forget to trigger the refresh. We don't yet have anything errors that would signify that one is working with outdated routing info but I imagine that instead of `TargetVersion::Latest`, the caller might specify something they got from a peer, such as monotonic a partition mapping version/epoch number. But the expectation is that whoever calls this already encountered an unrecoverable error, and calling this is just a nice to have signal. Let me see if I can work this into a doc comment.

Reply:
What do I do after I called this method? Right now, I cannot wait until the update has been processed. So I can only continue and hope that by the time I retry, the information has been updated. If not, then I'll talk to the outdated leader again.

Reply:
I was imagining the caller having a number of retries. Whenever it receives a "wrong node" response indicating that it is talking to a wrong node, it would like to refresh the routing information hoping that it was operating against wrong information. Once new information is obtained, it could continue hoping that this time the information is up to date.

If the caller stopped retrying, then only notifying the `RoutingInformation` that the data is outdated is probably enough.

Reply:
@tillrohrmann I think this is the main outstanding concern with this PR - would you like to see some specific improvements before we can merge this? I thought about this a fair amount yesterday and don't believe we can meaningfully improve on "just back off and retry" at the moment, though we'll definitely be able to do better in the future once we have better view of cluster node status.

Reply:
We can take this as a follow-up. One idea could be to have a watch that gets updated after the refresh task completes. A caller could then await this update, for example.

Reply:
Sounds good - I'd love to do that follow-up!